include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,373 @@
+    
+    void seed_internal(::std::initializer_list<scalar_type> seed) {
+        auto start = seed.begin();
+        auto end = seed.end();
+        // all counters are set to zero
+        for(::std::size_t i = 0; i < word_count; i++) {
+            state_.X[i] = 0;
+        }
+        // keys are set as seed
+        for (::std::size_t i = 0; i < (word_count/2); i++) {
+            state_.K[i] = (start == end) ? 0 : (*start++) & in_mask;
+        }
+        // results are set to zero
+        for (::std::size_t i = 0; i < word_count; i++) {
+            state_.Y[i] = 0;
+        }
+
+        state_.idx = word_count;
+    }
+    
+    /* Increment counter by 1 */
+    void increase_counter_internal() {
+        state_.X[0] = (state_.X[0] + 1) & in_mask;
+        for (::std::size_t i = 1; i < word_count; ++i) {
+            if (state_.X[i - 1]) {
+                [[likely]] return;
+            }
+            state_.X[i] = (state_.X[i] + 1) & in_mask;
+        }
+    }
+
+    /* Increment counter by an arbitrary z */
+    void increase_counter_internal(unsigned long long z) {
+        unsigned long long carry = 0;
+        unsigned long long ctr_inc = z;
+
+        for (::std::size_t i = 0; i < word_count; ++i) {
+            scalar_type initial_x = state_.X[i];
+            state_.X[i] = (initial_x + ctr_inc + carry) & in_mask;
+
+            carry = 0;
+            // check overflow of the current chunk
+            if(state_.X[i] < initial_x) {
+                carry = 1;
+            }
+
+            //          select high chunk            shift for addition with the next counter chunk
+            ctr_inc = (ctr_inc & (~in_mask)) >> (std::numeric_limits<unsigned long long>::digits - word_size);
+        }
+    }
+    
+    /* operator() specified for sycl_vec output */
+    template <unsigned int _N = 0>
+    ::std::enable_if_t<(_N > 0), result_type>
+    operator()() {
+        result_type tmp;
+        scalar_type curr_idx = state_.idx;
+
+        for (int elm_count = 0; elm_count < _N; elm_count++, curr_idx++) {
+            if(curr_idx  == word_count) { // empty buffer
+                generate();
+                increase_counter_internal();
+                curr_idx = 0;
+            }
+            tmp[elm_count] = state_.Y[curr_idx];
+        }
+        
+        state_.idx = ++curr_idx;
+
+        return tmp;
+    }
+    
+    /* operator() specified for a scalar output */
+    template <unsigned int _N = 0>
+    ::std::enable_if_t<(_N == 0), result_type>
+    operator()() {
+        result_type tmp;
+
+        scalar_type curr_idx = state_.idx;
+        if(curr_idx  == word_count) { // empty buffer
+            generate();
+            increase_counter_internal();
+            curr_idx = 0;
+        }
+        
+        // There are already generated numbers in the buffer
+        tmp = state_.Y[curr_idx];
+        state_.idx = ++curr_idx;
+
+        return tmp;
+    }
+
+public:
```

Comment:
Question: why private methods described first?

Reply:
Hmm, I didn't pay a lot of attention to the order of public/private members inside the class. If it's important I can reorder them.

Reply:
Just a question - please decide what is better here

Reply:
I took a look at oneMKL and llvm implementations - in oneMKL all public methods are specified first in all engine and distribution classes. In llvm the order of public and private members is a bit mixed, but it also tends to be "public at the top". So I also moved all public methods first.